src/modules/bulletproofs/tests_impl.h

```diff
@@ -0,0 +1,369 @@
+
+    secp256k1_weighted_scalar_inner_product(&res, a_vec, 0, a_vec, 0, 1, 4, &q);
+    secp256k1_scalar_set_int(&res2, 4740); /*i*i*4^(i+1) */
+    CHECK(secp256k1_scalar_eq(&res2, &res) == 1);
+
+    secp256k1_bulletproofs_powers_of_r(r_pows, &r, 4);
+    secp256k1_scalar_set_int(&res, 2); CHECK(secp256k1_scalar_eq(&res, &r_pows[0]));
+    secp256k1_scalar_set_int(&res, 4); CHECK(secp256k1_scalar_eq(&res, &r_pows[1]));
+    secp256k1_scalar_set_int(&res, 16); CHECK(secp256k1_scalar_eq(&res, &r_pows[2]));
+    secp256k1_scalar_set_int(&res, 256); CHECK(secp256k1_scalar_eq(&res, &r_pows[3]));
+}
+
+/* A complete norm argument. In contrast to secp256k1_bulletproofs_pp_rangeproof_norm_product_prove, this is meant
+   to be used as a standalone norm argument.
+   This is a simple wrapper around secp256k1_bulletproofs_pp_rangeproof_norm_product_prove
+   that also commits to the initial public values used in the protocol. In this case, these public
+   values are commitment.
+*/
+static int secp256k1_norm_arg_prove(
+    secp256k1_scratch_space* scratch,
+    unsigned char* proof,
+    size_t *proof_len,
+    const secp256k1_scalar* r_ch,
+    const secp256k1_bulletproofs_generators* g_vec,
+    const secp256k1_scalar* n_vec,
+    size_t n_vec_len,
+    const secp256k1_scalar* l_vec,
+    size_t l_vec_len,
+    const secp256k1_scalar* c_vec,
+    size_t c_vec_len,
+    const secp256k1_ge* commit
+) {
+    secp256k1_scalar *ns, *ls, *cs, r = *r_ch;
+    secp256k1_ge *gs, comm = *commit;
+    size_t scratch_checkpoint;
+    size_t g_len = n_vec_len, h_len = l_vec_len, i;
+    int res;
+    secp256k1_sha256 transcript;
+    unsigned char ser_commit[33];
+
+    scratch_checkpoint = secp256k1_scratch_checkpoint(&ctx->error_callback, scratch);
+    ns = (secp256k1_scalar*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, g_len * sizeof(secp256k1_scalar));
+    ls = (secp256k1_scalar*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, h_len * sizeof(secp256k1_scalar));
+    cs = (secp256k1_scalar*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, h_len * sizeof(secp256k1_scalar));
+    gs = (secp256k1_ge*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, (g_len + h_len) * sizeof(secp256k1_ge));
+    if (ns == NULL || ls == NULL || cs == NULL || gs == NULL) {
+        secp256k1_scratch_apply_checkpoint(&ctx->error_callback, scratch, scratch_checkpoint);
+        return 0;
+    }
+    memcpy(ns, n_vec, g_len * sizeof(secp256k1_scalar));
+    memcpy(ls, l_vec, h_len * sizeof(secp256k1_scalar));
+    memcpy(cs, c_vec, h_len * sizeof(secp256k1_scalar));
+    memcpy(gs, g_vec->gens, (g_len + h_len) * sizeof(secp256k1_ge));
+
+    /* Commit to the initial public values */
+    secp256k1_bulletproofs_pp_sha256_tagged_commitment_init(&transcript);
+    secp256k1_fe_normalize(&comm.x);
+    secp256k1_fe_normalize(&comm.y);
+    secp256k1_bulletproofs_serialize_pt(&ser_commit[0], &comm);
+    secp256k1_sha256_write(&transcript, ser_commit, 33);
```

Comment:
I think just using sha256_write to append to the transcript is dangerous. For example, if the transcript `T` consists of two vectors of scalars `l` and `r` such that:
```
l = [1, 2]; r = [3]; T = [l, r]
```
and similarly for transcript T' 
```
l' = [1]; r' = [2,3]; T' = [l', r']
```
Then `T != T'` but the challenge we compute is the same because input to SHA256 is just `[1,2,3]`. So it would be much better to have a transcript object that can be written to like sha256, but internally prepends the length (in bytes) of the input before calling sha256.

Reply:
Good catch. We should certainly prepend the length. I take it that you are suggesting that we serialize all the vector and include the total byte size as an argument.

I think it might be more logical to add a method to add another method to `transcript_consume_vec` that would add the len too. Obviously, this is footgunny because the user can write the for loop (just like I did above). But the approach where we push the entire vector as a blob feels odd. It also does not solve the above issue you mention where the user can still call items in a loop as individual elements.